src/sd.cpp

```diff
@@ -1091,36 +1093,146 @@ inline static std::vector<std::pair<std::string, float>> parse_prompt_attention(
+            [&tokenizer_bperankings](const auto& pair1, const auto& pair2)
+            {
+                int a = INT_MAX;
+                if (tokenizer_bperankings.find(pair1) != tokenizer_bperankings.end()) 
+                    a = tokenizer_bperankings[pair1];
+
+                int b = INT_MAX;
+                if (tokenizer_bperankings.find(pair2) != tokenizer_bperankings.end()) 
+                    b = tokenizer_bperankings[pair2];
 
-        if (pos < str.size())
+                return a < b;
+            }
+        );
+
+        if (bigram == pairs.end() || tokenizer_bperankings.find(*bigram) == tokenizer_bperankings.end())
+            break;
+
+        auto [first, second] = *bigram;
+
+        std::vector<std::string> new_word;
+        size_t i = 0;
+
+        while (i < word.size())
         {
-            std::string s = str.substr(i, pos - i);
-            std::string pat = std::string(1, str[pos]);
+            auto a = word.begin() + i;
+            if (a >= word.end())
+                a = word.end();
+
+            auto find_iter = std::find(word.begin()+i, word.end(), first);
+            if (find_iter == word.end())
+            {
+                new_word.insert(new_word.end(), a, word.end());
+                break;
+            }
+            else
+            {
+                size_t j = std::distance(word.begin(), find_iter);
+                auto b = word.begin() + j;
+                if (b >= word.end())
+                    b = word.end();
+                new_word.insert(new_word.end(), a, b);
+                i = j;
+            }
+
+            if (word[i] == first && i < word.size() - 1 && word[i + 1] == second) 
+            {
+                new_word.push_back(first + second);
+                i += 2;
+            }
+            else
+            {
+                new_word.push_back(word[i]);
+                i += 1;
+            }
+
+        }
+
+        // for (auto x : new_word) {
+        //     std::cout << x << " ";
+        // }
+        // std::cout << std::endl;;
 
+        word = new_word;
+
+        if (word.size() == 1)
+            break;
+        else
+            pairs = get_pairs(word);
+    }
+    return word;
+}
+
+inline static std::vector<std::string> split(std::string str, std::unordered_map<std::pair<std::string, std::string>, int, pair_hash>& tokenizer_bperankings)
+{
+    std::string const delims { "., " };
+    
+    size_t beg, pos = 0;
+    std::vector<std::string> result;
+    while ((beg = str.find_first_not_of(delims, pos)) != std::string::npos)
+    {
+        pos = str.find_first_of(delims, beg + 1);
+        std::string s = str.substr(beg, pos - beg);
+
+        if (g_main_args.m_bpe_ranking)
+        {
+            std::vector<std::string> bpes = bpe(s, tokenizer_bperankings);
+            result.insert(result.end(), bpes.begin(), bpes.end());
+        }
+        else 
+        {
+            std::string pat = std::string(1, str[pos]);
```

Comment:
This original algorithm had issues with things like a comma then a space, so `, ` and also seemed to have issues with `.`
 I replaced it with one that seems to do the trick